lib/doctor.js

```diff
@@ -1,115 +1,294 @@
+const isWindows = require('./utils/is-windows.js')
+const checkFilesPermission = async (root, shouldOwn = true, mask = null) => {
+  if (mask === null) {
+    mask = shouldOwn ? R_OK | W_OK : R_OK
   }
 
-  const actionsToRun = [
-    [checkPing],
-    [getLatestNpmVersion],
-    [getLatestNodejsVersion, args['node-url']],
-    [getGitPath],
-    [checkFilesPermission, npm.cache, 4, 6],
-    [checkFilesPermission, globalNodeModules, 4, 4],
-    [checkFilesPermission, localNodeModules, 6, 6],
-    [verifyCachedFiles, path.join(npm.cache, '_cacache')]
-  ]
+  let ok = true
+
+  const tracker = npm.log.newItem(root, 1)
+
+  try {
+    const uid = process.getuid()
+    const gid = process.getgid()
+    const files = new Set([root])
+    for (const f of files) {
+      tracker.silly('checkFilesPermission', f.substr(root.length + 1))
+      const st = await lstat(f)
+        .catch(er => {
+          ok = false
+          tracker.warn('checkFilesPermission', 'error getting info for ' + f)
+        })
+
+      tracker.completeWork(1)
 
-  log.info('doctor', 'Running checkup')
-  chain(actionsToRun, function (stderr, stdout) {
-    if (stderr && stderr.message !== 'not found: git') return cb(stderr)
-    const list = makePretty(stdout)
-    let outHead = ['Check', 'Value', 'Recommendation']
-    let outBody = list
-
-    if (npm.color) {
-      outHead = outHead.map(function (item) {
-        return styles.underline(item)
-      })
-      outBody = outBody.map(function (item) {
-        if (item[2]) {
-          item[0] = color.red(item[0])
-          item[2] = color.magenta(item[2])
+      if (!st) {
+        continue
+      }
+
+      if (shouldOwn && (uid !== st.uid || gid !== st.gid)) {
+        tracker.warn('checkFilesPermission', 'should be owner of ' + f)
+        ok = false
+      }
+
+      if (!st.isDirectory() && !st.isFile()) {
+        continue
+      }
+
+      try {
+        await access(f, mask)
+      } catch (er) {
+        ok = false
+        const msg = `Missing permissions on ${f} (expect: ${maskLabel(mask)})`
+        tracker.error('checkFilesPermission', msg)
+        continue
+      }
+
+      if (st.isDirectory()) {
+        const entries = await readdir(f)
+          .catch(er => {
+            ok = false
+            tracker.warn('checkFilesPermission', 'error reading directory ' + f)
+            return []
+          })
+        for (const entry of entries) {
+          files.add(resolve(f, entry))
         }
-        return item
-      })
+      }
     }
-
-    const outTable = [outHead].concat(outBody)
-    const tableOpts = {
-      stringLength: function (s) { return ansiTrim(s).length }
+  } finally {
+    tracker.finish()
+    if (!ok) {
+      throw `Check the permissions of files in ${root}` +
+        (shouldOwn ? ' (should be owned by current user)' : '')
+    } else {
+      return ''
     }
+  }
+}
 
-    if (!silent) output(table(outTable, tableOpts))
+const which = promisify(require('which'))
```

Comment:
isn't the `which` API promisified already in v2?

Reply:
Oh!  So it is.  It still accepts a cb, which is what threw me.  I'll update and push.